diff --git a/lib/server_connection.ml b/lib/server_connection.ml
index 44854e4..a252c41 100644
--- a/lib/server_connection.ml
+++ b/lib/server_connection.ml
@@ -618,7 +618,9 @@ let process_data_frame t { Frame.frame_header; _ } bstr =
      *   unless the receiver treats this as a connection error (Section 5.4.1).
      *   This is necessary even if the frame is in error. *)
     let payload_len32 = Int32.of_int payload_length in
-    Scheduler.deduct_inflow t.streams payload_len32;
+    let deduct_connection_inflow () =
+      Scheduler.deduct_inflow t.streams payload_len32
+    in
     match Scheduler.get_node t.streams stream_id with
     | Some (Stream { descriptor; _ } as stream) ->
       (match descriptor.state with
@@ -630,7 +632,22 @@ let process_data_frame t { Frame.frame_header; _ } bstr =
               request_info.request_body_bytes
               (of_int (Bigstringaf.length bstr)));
         let request = request_info.request in
-        if not Scheduler.(allowed_to_receive t.streams stream payload_len32)
+        (* First, calculate whether we're allowed to receive this frame based
+         * on the current inflow. *)
+        let allowed_to_receive =
+          Scheduler.(allowed_to_receive t.streams stream payload_len32)
+        in
+        (* Then, deduct inflow from the connection flow-control window, as
+         * mandated by the protocol.
+         *
+         * From RFC7540§6.9:
+         *   A receiver that receives a flow-controlled frame MUST always
+         *   account for its contribution against the connection flow-control
+         *   window, unless the receiver treats this as a connection error
+         *   (Section 5.4.1). This is necessary even if the frame is in
+         *   error. *)
+        deduct_connection_inflow ();
+        if not allowed_to_receive
         then (
           (* From RFC7540§6.9:
            *  A receiver MAY respond with a stream error (Section 5.4.2) or
@@ -699,6 +716,7 @@ let process_data_frame t { Frame.frame_header; _ } bstr =
          *   window, unless the receiver treats this as a connection error
          *   (Section 5.4.1). This is necessary even if the frame is in
          *   error. *)
+        deduct_connection_inflow ();
         send_window_update t t.streams payload_len32
       (* From RFC7540§6.4:
        *   [...] after sending the RST_STREAM, the sending endpoint MUST be
@@ -711,6 +729,7 @@ let process_data_frame t { Frame.frame_header; _ } bstr =
        * as a way of only accepting frames after an RST_STREAM from us up to
        * a time limit. *)
       | _ ->
+        deduct_connection_inflow ();
         send_window_update t t.streams payload_len32;
         (* From RFC7540§6.1:
          *   If a DATA frame is received whose stream is not in "open" or
@@ -725,6 +744,7 @@ let process_data_frame t { Frame.frame_header; _ } bstr =
          *   a stream in this state MUST be treated as a connection error
          *   (Section 5.4.1) of type PROTOCOL_ERROR. *)
         report_connection_error t Error_code.ProtocolError
+      else deduct_connection_inflow ()

 let process_priority_frame t { Frame.frame_header; _ } priority =
   let { Frame.stream_id; _ } = frame_header in
